webapp/src/main/java/com/box/l10n/mojito/smartling/SmartlingClient.java

```diff
@@ -0,0 +1,199 @@
+import com.box.l10n.mojito.smartling.response.SourceStringsResponse;
+import com.ibm.icu.text.MessageFormat;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.*;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriTemplate;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.*;
+
+/**
+ * Smartling client based on the Web API: https://github.com/Smartling/api-sdk-java
+ */
+public class SmartlingClient {
+
+    /**
+     * logger
+     */
+    static Logger logger = LoggerFactory.getLogger(SmartlingClient.class);
+
+    static final String DEFAULT_BASE_HOST = "https://api.smartling.com/";
+    static final String AUTH_API_V2_AUTHENTICATE = DEFAULT_BASE_HOST + "auth-api/v2/authenticate";
+    static final String AUTH_API_V2_REFRESH = DEFAULT_BASE_HOST + "auth-api/v2/authenticate/refresh";
+    static final String API_PULL_SOURCE_STRINGS = DEFAULT_BASE_HOST + "strings-api/v2/projects/{projectId}/source-strings?fileUri={fileUri}&offset={offset}";
+    static final String API_PUSH_NEW_CONTEXT = DEFAULT_BASE_HOST + "context-api/v2/projects/{projectId}/contexts";
+    static final String API_BIND_CONTEXT_TO_HASHCODE = DEFAULT_BASE_HOST + "context-api/v2/projects/{projecId}/bindings";
+
+    RestTemplate restTemplate = new RestTemplate();
+
+    String userIdentifier;
+    String userSecret;
+    String authToken = null;
+    String refreshToken = null;
+    Instant lastAuthentication = null;
+    Instant lastRefresh = null;
+    Instant nextAuthentication = null;
+    Instant nextRefresh = null;
+    // Refresh tokens are valid for up to 1 hour
+    Integer refreshDurationMinutes = 50;
+    // Authentication tokens can be refreshed for up to 24 hours
+    Integer authenticationDurationHours = 23;
+
+    public SmartlingClient(String userIdentifier, String userSecret) {
+        this.userIdentifier = userIdentifier;
+        this.userSecret = userSecret;
+    }
+
+    public enum RequestType {
+        AUTHENTICATION,
+        REFRESH,
+        SOURCE_STRINGS,
+        WITH_TOKEN
+    }
+
+    Map<String, Object> getPayloadMap(RequestType requestType) {
+        Map<String, Object> map = new HashMap<>();
+        if (requestType == RequestType.AUTHENTICATION) {
+            map.put("userIdentifier", userIdentifier);
+            map.put("userSecret", userSecret);
+        } else if (requestType == RequestType.REFRESH) {
+            map.put("refreshToken", refreshToken);
+        }
+        return map;
+    }
+
+    HttpHeaders getHttpHeadersForJson(RequestType requestType) {
+        HttpHeaders headers = new HttpHeaders();
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        if (requestType == RequestType.SOURCE_STRINGS) {
+            headers.set("Authorization", "Bearer " + authToken);
```

Comment:
why do we set only the auth header for "source string", looks like we should have a function for unauth and one for auth request. So following previous comment, probably better to split in 2 functions.

The logic for unauth is probably gonna be unique and might be inlined. vs auth that will be shared by all API calls

Reply:
we set no auth header for auth and refresh because they are not required, I don't think it needs to be split https://api-reference.smartling.com/#tag/Authentication

Reply:
my comment was unclear. I got that it doesn't set the header since it is fetching the information for it. 

what I was trying to say is:
1) if more APIs get added to the client, for each of those, the "if" statement will have to be updated with the new request type and that doesn't scale. It is also hard to test and overall won't be clean
2) there is no point during the auth flow to call a function that has logic specific to the API calls that need the authorization header. 

So there should be a base method that gets the header for JSON (used by the auth flow), then have a method that gets header for requests that needs authentication (get source string and new ones later). The later method calls the method to get header for JSON and then adds the authorization header. 

And this was related to the previous comment. there shouldn't be a single `getPayloadMap` method with a if statement for each request type. Instead there should be a function for each request type.